File: java/shared/java/org/signal/libsignal/protocol/kem/KEMPublicKey.java
//
// Copyright 2023 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.signal.libsignal.protocol.kem;

import static org.signal.libsignal.internal.FilterExceptions.filterExceptions;

import java.util.Arrays;
import org.signal.libsignal.internal.Native;
import org.signal.libsignal.internal.NativeHandleGuard;
import org.signal.libsignal.protocol.InvalidKeyException;
import org.signal.libsignal.protocol.SerializablePublicKey;

public class KEMPublicKey extends NativeHandleGuard.SimpleOwner implements SerializablePublicKey {

  public KEMPublicKey(byte[] serialized, int offset) throws InvalidKeyException {
    super(
        filterExceptions(
            InvalidKeyException.class,
            () -> Native.KyberPublicKey_DeserializeWithOffset(serialized, offset)));
  }

  public KEMPublicKey(byte[] serialized) throws InvalidKeyException {
    super(
        filterExceptions(
            InvalidKeyException.class,
            () -> Native.KyberPublicKey_DeserializeWithOffset(serialized, 0)));
  }

  public KEMPublicKey(long nativeHandle) {
    super(KEMPublicKey.throwIfNull(nativeHandle));
  }

  private static long throwIfNull(long handle) {
    if (handle == 0) {
      throw new NullPointerException();
    }
    return handle;
  }

  @Override
  protected void release(long nativeHandle) {
    Native.KyberPublicKey_Destroy(nativeHandle);
  }

  public byte[] serialize() {
    return filterExceptions(() -> guardedMapChecked(Native::KyberPublicKey_Serialize));
  }

  @Override
  public boolean equals(Object other) {
    if (other == null) return false;
    if (!(other instanceof KEMPublicKey)) return false;
    return guardedMap(
        (thisNativeHandle) ->
            ((KEMPublicKey) other)
                .guardedMap(
                    (otherNativeHandle) ->
                        Native.KyberPublicKey_Equals(thisNativeHandle, otherNativeHandle)));
  }

  @Override
  public int hashCode() {
    return Arrays.hashCode(this.serialize());
  }
}

File: java/shared/java/org/signal/libsignal/protocol/kem/KEMPublicKey.kt
//
// Copyright 2023 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.signal.libsignal.protocol.kem

import org.signal.libsignal.internal.Native
import org.signal.libsignal.internal.NativeHandleGuard
import org.signal.libsignal.protocol.InvalidKeyException
import org.signal.libsignal.protocol.SerializablePublicKey
import java.util.Arrays

public class KEMPublicKey :
  NativeHandleGuard.SimpleOwner,
  SerializablePublicKey {
  @Deprecated("use the constructor that takes an offset and length")
  @Throws(InvalidKeyException::class)
  public constructor(serialized: ByteArray, offset: Int) :
    this(serialized, offset, length = serialized.size - offset)

  @Throws(InvalidKeyException::class)
  public constructor(serialized: ByteArray, offset: Int, length: Int) :
    super(Native.KyberPublicKey_DeserializeWithOffsetLength(serialized, offset, length))

  @Throws(InvalidKeyException::class)
  public constructor(serialized: ByteArray) : this(serialized, 0, serialized.size)

  public constructor(nativeHandle: Long) : super(NativeHandleGuard.SimpleOwner.throwIfNull(nativeHandle))

  protected override fun release(nativeHandle: Long) {
    Native.KyberPublicKey_Destroy(nativeHandle)
  }

  public fun serialize(): ByteArray = guardedMapChecked(Native::KyberPublicKey_Serialize)

  public override fun equals(other: Any?): Boolean =
    when (other) {
      null -> false
      is KEMPublicKey ->
        guardedMap { thisNativeHandle ->
          other.guardedMap { otherNativeHandle ->
            Native.KyberPublicKey_Equals(thisNativeHandle, otherNativeHandle)
          }
        }
      else -> false
    }

  public override fun hashCode(): Int = Arrays.hashCode(this.serialize())
}
